File: sdk/dart/hejaz_ai_sdk/lib/src/exceptions.dart
/// Exception hierarchy for the Hejaz AI SDK.
///
/// All exceptions carry structured metadata (statusCode, message, requestId)
/// so callers can log, retry, or surface errors without parsing strings.

/// Base exception for all SDK errors.
class HejazAIException implements Exception {
  /// Human-readable error message.
  final String message;

  /// HTTP status code, if the error originated from an HTTP response.
  final int? statusCode;

  /// Server-assigned request ID for correlation, if available.
  final String? requestId;

  const HejazAIException(
    this.message, {
    this.statusCode,
    this.requestId,
  });

  @override
  String toString() {
    final parts = <String>['HejazAIException: $message'];
    if (statusCode != null) parts.add('statusCode=$statusCode');
    if (requestId != null) parts.add('requestId=$requestId');
    return parts.join(', ');
  }
}

/// Authentication or authorisation failure (HTTP 401 / 403).
class AuthException extends HejazAIException {
  const AuthException(
    super.message, {
    super.statusCode,
    super.requestId,
  });

  @override
  String toString() {
    final parts = <String>['AuthException: $message'];
    if (statusCode != null) parts.add('statusCode=$statusCode');
    if (requestId != null) parts.add('requestId=$requestId');
    return parts.join(', ');
  }
}

/// Too many requests (HTTP 429).
class RateLimitException extends HejazAIException {
  /// Seconds the server suggests waiting before retrying, if provided.
  final double? retryAfter;

  const RateLimitException(
    super.message, {
    super.statusCode = 429,
    super.requestId,
    this.retryAfter,
  });

  @override
  String toString() {
    final parts = <String>['RateLimitException: $message'];
    if (statusCode != null) parts.add('statusCode=$statusCode');
    if (retryAfter != null) parts.add('retryAfter=${retryAfter}s');
    if (requestId != null) parts.add('requestId=$requestId');
    return parts.join(', ');
  }
}

/// Requested resource does not exist (HTTP 404).
class NotFoundException extends HejazAIException {
  const NotFoundException(
    super.message, {
    super.statusCode = 404,
    super.requestId,
  });
}

/// Request payload failed server-side validation (HTTP 422).
class ValidationException extends HejazAIException {
  const ValidationException(
    super.message, {
    super.statusCode = 422,
    super.requestId,
  });
}

/// Internal server error (HTTP 5xx).
class ServerException extends HejazAIException {
  const ServerException(
    super.message, {
    super.statusCode,
    super.requestId,
  });
}

/// SSE connection or parsing failure during streaming.
class StreamException extends HejazAIException {
  const StreamException(
    super.message, {
    super.statusCode,
    super.requestId,
  });
}

/// Request exceeded the configured timeout.
class TimeoutException extends HejazAIException {
  const TimeoutException(
    super.message, {
    super.statusCode,
    super.requestId,
  });
}

/// Map an HTTP status code to the appropriate exception type.
HejazAIException exceptionFromStatus(
  int statusCode,
  String message, {
  String? requestId,
  double? retryAfter,
}) {
  switch (statusCode) {
    case 401:
    case 403:
      return AuthException(message,
          statusCode: statusCode, requestId: requestId);
    case 404:
      return NotFoundException(message,
          statusCode: statusCode, requestId: requestId);
    case 422:
      return ValidationException(message,
          statusCode: statusCode, requestId: requestId);
    case 429:
      return RateLimitException(message,
          statusCode: statusCode,
          requestId: requestId,
          retryAfter: retryAfter);
    case >= 500:
      return ServerException(message,
          statusCode: statusCode, requestId: requestId);
    default:
      return HejazAIException(message,
          statusCode: statusCode, requestId: requestId);
  }
}

File: sdk/python/ai_assistant/exceptions.py
"""
Exception hierarchy for the AI Assistant SDK.

All exceptions carry structured metadata (status_code, message, request_id)
so callers can log, retry, or surface errors without parsing strings.
"""

from __future__ import annotations


class AssistantError(Exception):
    """Base exception for all SDK errors."""

    def __init__(
        self,
        message: str,
        *,
        status_code: int | None = None,
        request_id: str | None = None,
    ) -> None:
        super().__init__(message)
        self.message = message
        self.status_code = status_code
        self.request_id = request_id

    def __repr__(self) -> str:
        parts = [f"message={self.message!r}"]
        if self.status_code is not None:
            parts.append(f"status_code={self.status_code}")
        if self.request_id is not None:
            parts.append(f"request_id={self.request_id!r}")
        return f"{type(self).__name__}({', '.join(parts)})"


class AuthError(AssistantError):
    """Authentication or authorization failure (HTTP 401 / 403)."""


class NotFoundError(AssistantError):
    """Requested resource does not exist (HTTP 404)."""


class RateLimitError(AssistantError):
    """Too many requests (HTTP 429).

    Attributes:
        retry_after: Seconds the server suggests waiting before retrying.
    """

    def __init__(
        self,
        message: str,
        *,
        status_code: int | None = 429,
        request_id: str | None = None,
        retry_after: float | None = None,
    ) -> None:
        super().__init__(message, status_code=status_code, request_id=request_id)
        self.retry_after = retry_after


class ValidationError(AssistantError):
    """Request payload failed server-side validation (HTTP 422)."""


class ServerError(AssistantError):
    """Internal server error (HTTP 5xx)."""


class StreamError(AssistantError):
    """SSE connection or parsing failure."""


class TimeoutError(AssistantError):  # noqa: A001 — intentionally shadows builtins.TimeoutError
    """Request exceeded the configured timeout."""


# ---------------------------------------------------------------------------
# Mapping helpers (used by transport layer)
# ---------------------------------------------------------------------------

_STATUS_MAP: dict[int, type[AssistantError]] = {
    401: AuthError,
    403: AuthError,
    404: NotFoundError,
    422: ValidationError,
    429: RateLimitError,
    500: ServerError,
    502: ServerError,
    503: ServerError,
    504: ServerError,
}


def error_from_status(
    status_code: int,
    message: str,
    *,
    request_id: str | None = None,
    retry_after: float | None = None,
) -> AssistantError:
    """Instantiate the appropriate exception for an HTTP status code."""
    cls = _STATUS_MAP.get(status_code, AssistantError)
    if cls is RateLimitError:
        return RateLimitError(
            message,
            status_code=status_code,
            request_id=request_id,
            retry_after=retry_after,
        )
    return cls(message, status_code=status_code, request_id=request_id)
